lib/authHttpClient.js

```
@@ -0,0 +1,72 @@
+import bedrock from 'bedrock';
+const {BedrockError} = bedrock;
+
+const proxyMethods = new Set([
+  'get', 'post', 'push', 'patch', 'head', 'delete'
+]);
+
+export function createAuthzHttpClient({oAuth2Client} = {}) {
+  const authHttpClient = new Proxy(httpClient, {
+    async apply(target, thisArg, args) {
+      return _handleRequest(target, this, args, oAuth2Client);
+    },
+    get(target, propKey) {
+      const propValue = target[propKey];
+
+      // only intercept particular methods
+      if(!proxyMethods.has(propKey)) {
+        return propValue;
+      }
+      return async function() {
+        return _handleRequest(propValue, this, arguments, oAuth2Client);
+      };
+    }
+  });
+  return authHttpClient;
+}
+
+async function _handleRequest(target, thisArg, args, oAuth2Client) {
+  const accessToken = await getAccessToken({...oAuth2Client});
+
+  // create the authz header using the token
+  // add the authz header to the headers object of args
+  const [url, options] = args;
+  options.headers = options.headers || {};
+  if(options.headers.Authorization) {
+    options.headers.Authorization = [options.headers.Authorization];
+    options.headers.Authorization.push(`Bearer ${accessToken}`);
+  } else {
+    options.headers.Authorization = `Bearer ${accessToken}`;
+  }
+  let response;
+  try {
+    response = await target.apply(thisArg, [url, options]);
+  } catch(error) {
+    return _handleTokenError({
+      error, target, thisArg, url, options, oAuth2Client
+    });
+  }
+  return response;
+}
+
+async function _handleTokenError({
+  error, target, thisArg, url, options, oAuth2Client
+}) {
+  if(!isInvalidAccessTokenError({error})) {
+    throw error;
+  }
+  const accessToken = await getNewAccessToken({...oAuth2Client, maxRetries: 3});
+  if(accessToken) {
+    return target.apply(thisArg, [url, options]);
+  }
+  throw new BedrockError('Could not get access token.', 'DataError', {
```

Actually, I don't think this code path is ever hit. If `getNewAccessToken` fails after `maxRetries`, it throws. There's no case where it returns an empty/invalid `accessToken`, right? If so, we should just let it throw and we don't actually care about the return value.

However, looking at this, it seems like it would retry POST operations (not just GET / idempotent operations). That seems like it could be a bad idea. Were we previously automatically retrying POST operations when a token would fail?

---

@dlongley The immediate intended use of the library is to re-try POST operations, and that was previously occurring.

---

Ok, since this has been previously vetted and that's the main purpose, I'll resolve this particular thread.